### PR TITLE
pkg/openwsn: re-integrate the network stack as a package

### PR DESCRIPTION
**Pull request by PeterKietzmann**
### Contribution description
This PR re-integrates the OpenWSN network stack as a pkg to RIOT. It is marked as WIP because it does not yet work stable and I'm asking for your help in terms of debugging, testing and review.

An example application is provided, including documentation in the [README](https://github.com/RIOT-OS/RIOT/pull/8570/files#diff-af628c95ba6f73b8a95c7f7993036f63) next to it. Furthermore, I added [documentation](https://github.com/RIOT-OS/RIOT/pull/8570/files#diff-71c35b7e125127740270b42094c8dd75) about the port in general, the status of this implementation as well as additional information about debugging capabilities and currently open issues/TODOs. The two commits covering Atmel radio driver changes and iotlab-m3 timer configuration are not intended to be merged but rather to be included as a patch later.

I have a collection of additional test applications for different wrappers lying around, which I would provide in a separate PR if we agree on how to integrate them into the repository.

### Implementation Issues 
~~- Nodes reset after some time ([here?](https://github.com/PeterKietzmann/openwsn-fw/blob/RIOT_port/kernel/openos/scheduler.c#L98)). **Without** channel hopping this effect does not show up that often. **With** channel hopping it gets worse~~
~~- After a node restarts, it won't join the topology any time soon, even though neighbored nodes still send DIOs~~

### People that might be interested and in charge to support
(alphabetical order of last name)
@emmanuelsearch, @kb2ma, @changtengfei, @OlegHahm, @sieben, @fjmolinas, @Lotterleben, @twatteyne

### Issues/PRs references
Relates to #3039 and #7893

### CODE REVIEW

**Comment by PeterKietzmann** [comment]
Useful hint no 1: For easier debugging I'm using a local copy of the adapted `openwsn-fw` repository. To incorporate the local repo I replace the `git-dowlnload` command in the pkg [Makefile](https://github.com/RIOT-OS/RIOT/pull/8570/files#diff-0997f22c3baa7840b772ffe4632b957aR9) by something like:
```
 OPENWSN_PATH := ../../../openwsn-fw
 all:
 	rm -rf $(PKG_BUILDDIR) && mkdir -p $(PKG_BUILDDIR)
 	rsync -a $(OPENWSN_PATH) $(dir $(PKG_BUILDDIR)) --exclude .git
 	$(MAKE) -C $(PKG_BUILDDIR)
```

**Comment by PeterKietzmann** [comment]
Useful hint no 2: RIOT provides some build commands to easily submit experiments to the FIT IoT.LAB testbed. For documentation see [here](https://github.com/RIOT-OS/RIOT/blob/master/dist/testbed-support/README.iotlab.md). TL;DR: From the test application directory use this command to build, submit an experiment with three nodes and start a tmux session to the aggregated serial output of that experiment:
`BOARD=iotlab-m3 make clean all iotlab-exp iotlab-term IOTLAB_NODES=3`

File: pkg/openwsn/contrib/eui64.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "string.h"
+#include "eui64.h"
+
+#include "luid.h"
+#include "net/ieee802154.h"
+
+void eui64_get(uint8_t *addressToWrite)
+{
+    luid_get((void *)addressToWrite, IEEE802154_LONG_ADDRESS_LEN);
```
Comment (miri64) [inline pkg/openwsn/contrib/eui64.c]: This will get you a different EUI-64 everytime the function is called or is this intended (couldn't find any documentation, on this but I saw it is called [twice here](https://github.com/openwsn-berkeley/openwsn-fw/blob/2aa270d1da7fb2bb573372d994d1e3cca041288b/openapps/cjoin/cjoin.c#L88-L90))?

If not, why not get the long-address from the device somehow? E.g. have

```C
netdev_t *openwsn_netdev;

void eui64_get(uint8_t *addressToWrite)
{
    eui64_t eui64;
    if (openwsn_netdev->driver->get(dev, NETOPT_ADDRESS_LONG, &eui64, sizeof(eui64_t)) == sizeof(eui64_t)) {
        memcpy(addressToWrite, eui64.u8, sizeof(eui64.u8));
    }
    else {
        /* Use your code as a fallback or print an error message here */
        luid_get((void *)addressToWrite, IEEE802154_LONG_ADDRESS_LEN);
    }
}
```

And set netdev in `radio_init()`:

```C
extern netdev_t *openwsn_netdev;

void radio_init(void)
{
    /* ... */
#ifdef MODULE_AT86RF2XX
    openwsn_netdev = (netdev_t *)&at86rf2xx_dev.netdev.netdev;
    radio_vars.dev = openwsn_netdev;
    at86rf2xx_setup(&at86rf2xx_dev, &at86rf2xx_params[0]);
#endif
```

If you fear memory waste, take `netdev` out of `radio_vars_t` and use `openwsn_netdev` instead of `radio_vars.dev`.

/cc @fjmolinas

File: boards/common/iotlab/include/periph_conf_common.h [hunk 2]
```diff
@@ -135,8 +135,8 @@ static const uart_conf_t uart_config[] = {
 #define RTT_IRQ             RTC_IRQn
 #define RTT_ISR             isr_rtc
 #define RTT_MAX_VALUE       (0xffffffff)
-#define RTT_FREQUENCY       (1)             /* in Hz */
-#define RTT_PRESCALER       (0x7fff)        /* run with 1 Hz */
+#define RTT_FREQUENCY       (16384U)     /* in Hz */
+#define RTT_PRESCALER       (0x1)        /* run with ~16 kHz */
```
Comment (miri64) [inline boards/common/iotlab/include/periph_conf_common.h]: Probably still a TODO, but these should be configured at compile-time, when OpenWSN is present.

File: drivers/at86rf2xx/at86rf2xx.c [hunk 3]
```diff
@@ -147,6 +147,9 @@ void at86rf2xx_tx_prepare(at86rf2xx_t *dev)
     if (state != AT86RF2XX_STATE_TX_ARET_ON) {
         dev->idle_state = state;
     }
+
+    at86rf2xx_set_state(dev, AT86RF2XX_STATE_PLL_ON);
```
Comment (miri64) [inline drivers/at86rf2xx/at86rf2xx.c]: Probably also a TODO, but hard-wiring states required for OpenWSN seems like a bad idea to me.

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 4]
```diff
@@ -213,7 +213,7 @@ static int _set_state(at86rf2xx_t *dev, netopt_state_t state)
                      * know when to switch back to the idle state. */
                     ++dev->pending_tx;
                 }
-                at86rf2xx_set_state(dev, AT86RF2XX_STATE_TX_ARET_ON);
+                at86rf2xx_set_state(dev, AT86RF2XX_STATE_PLL_ON);
```
Comment (miri64) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: Dito.

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 5]
```diff
@@ -660,6 +660,21 @@ static void _isr(netdev_t *netdev)
...
+            if (netdev->event_callback && (dev->netdev.flags & AT86RF2XX_OPT_TELL_TX_END)) {
+                if ((trac_status>>5)==7 || trac_status==AT86RF2XX_TRX_STATE__TRAC_SUCCESS) {
+                    /* Even though the reset value for register bits TRAC_STATUS is 0, the RX_AACK and
+                     * TX_ARET procedures set the register bits to TRAC_STATUS = 7 (INVALID) when it is started. */
+                    netdev->event_callback(netdev, NETDEV_EVENT_TX_COMPLETE);
+                    DEBUG("[at86rf2xx] TX SUCCESS\n");
+                }  else {
+                    DEBUG("[at86rf2xx] Unhandled TRAC_STATUS: %d\n",trac_status >> 5);
+                }
+            }
```
Comment (miri64) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: Move to separate PR?

File: pkg/openwsn/contrib/radio.c [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+static msg_t _queue[OW_NETDEV_QUEUE_LEN];
+
+typedef struct {
+    radio_capture_cbt startFrame_cb;
+    radio_capture_cbt endFrame_cb;
+    radio_state_t state;
+    netdev_t                 *dev;
+} radio_vars_t;
+
+radio_vars_t radio_vars;
```
Comment (miri64) [inline pkg/openwsn/contrib/radio.c]: Isn't exposed, so `static`?

**Comment by PeterKietzmann** [comment]
The above mentioned issues have been solved. Furthermore, I squashed and rebased both RIOT as well as openwsn. Help in form of reviewing and testing is happily welcome. There is still work TODO which can be taken from the [doc.txt](https://github.com/PeterKietzmann/RIOT/blob/OpenWSN_port/pkg/openwsn/doc.txt). Most importantly:
- find a generic way to enable the basic mode of the Atmel radio
- make the sctimer wrapper hardware independant
- run on different platforms with different radios

**Comment by miri64** [comment]
> find a generic way to enable the basic mode of the Atmel radio

Can https://github.com/RIOT-OS/RIOT/pull/9592 help with that?

**Comment by kb2ma** [comment]
Able to send a simple UDP packet on IoT-LAB nodes using the test application. It's a start! FWIW, unable to compile with CFLAGS=-DOW_RIOT_SINGLE_CHANNEL=26. Lots of errors. But if channel hopping works, that's fine with me.

Next I'll see if I can use OpenWSN as a sock-like provider for nanocoap.

> udp send bbbb::b85:fb65:106b:1115 61617 1234
udp send bbbb::b85:fb65:106b:1115 61617 1234
Send 4 byte over UDP to [bbbb::b85:fb65:106b:1115]:61617
> Send success

udp server start 61617
Set UDP server port to 61617
> Received 13 bytes on port 61617
00000000  45  81  7A  73  00  00  00  00  00  31  32  33  34              E.zs.....1234

**Comment by miri64** [comment]
Ping @PeterKietzmann?

**Comment by PeterKietzmann** [comment]
Thanks for the feedback, I'll address what I can. There was some "major" reconstruction and added features to this work which is why I'm closing the PR for now. Will re-open once ready.
